frame/session/src/historical/offchain.rs

```diff
@@ -0,0 +1,263 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2019-2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Off-chain logic for creating a proof based data provided by on-chain logic.
+//!
+//! Validator-set extracting an iterator from an off-chain worker stored list containing
+//! historical validator-sets.
+//! Based on the logic of historical slashing, but the validation is done off-chain.
+//! Use [`fn store_current_session_validator_set_to_offchain()`](super::onchain) to store the
+//! required data to the offchain validator set.
+//! This is used in conjunction with [`ProvingTrie`](super::ProvingTrie) and
+//! the off-chain indexing API.
+
+use sp_runtime::{offchain::storage::StorageValueRef, KeyTypeId};
+use sp_session::MembershipProof;
+
+use super::super::{Module as SessionModule, SessionIndex};
+use super::{IdentificationTuple, ProvingTrie, Trait};
+
+use super::shared;
+use sp_std::prelude::*;
+
+
+/// A set of validators, which was used for a fixed session index.
+struct ValidatorSet<T: Trait> {
+	validator_set: Vec<IdentificationTuple<T>>,
+}
+
+impl<T: Trait> ValidatorSet<T> {
+	/// Load the set of validators for a particular session index from the off-chain storage.
+	///
+	/// If none is found or decodable given `prefix` and `session`, it will return `None`.
+	/// Empty validator sets should only ever exist for genesis blocks.
+	pub fn load_from_offchain_db(session_index: SessionIndex) -> Option<Self> {
+		let derived_key = shared::derive_key(shared::PREFIX, session_index);
+		StorageValueRef::persistent(derived_key.as_ref())
```

Comment:
Shouldn't this be the unimplemented `StorageValueRef::local`  (in general got a hard time seing how this will work with reorg)?

Reply:
There was a discussion about re-orgs with @tomusdrw and @rphmeier and it turns out that the validator set per session is consistent across re-orgs regarding the session index.

Reply:
Yes, this is because sessions are buffered by 1